### PR TITLE
Add: [AI/GS] Missing water related functions and objects

### PR DESCRIPTION
**Pull request by SamuXarick**
Allows scripts to get the costs of building lock, building canal and clearing water (sea or river in this case).
Allows scripts to check whether a tile is sea and whether a tile is river (checking whether a tile is canal was already possible).

I didn't add the costs for clearing a lock, or canal. Maybe I should, but I didn't want to overdo it.

### CODE REVIEW

**Comment by glx22** [comment]
Maybe add regression checks for added AITile functions

**Comment by SamuXarick** [comment]
Unnamed, 1954-11-25

This is the new area where the water, river, canal, coast tests are being done for regression. The old area was too simple.

File: regression/regression/result.txt [hunk 1]
```diff
@@ -9099,12 +9272,12 @@ ERROR: IsEnd() is invalid as Begin() is never called
     GetLocation():       33417
     GetEngineType():     153
     GetUnitNumber():     1
-    GetAge():            0
+    GetAge():            1
```
Comment (LordAro) [inline regression/regression/result.txt:9275]: Why are these (and the following) apparently unrelated changes to do with age necessary?

Comment (nielsmh) [reply regression/regression/result.txt:9275]: A guess could be that the number of operations before the AI is suspended and another game tick simulated is reached with the new tests inserted.

**Review by glx22 — DISMISSED** [review]

File: src/script/api/script_tile.hpp [hunk 2]
```diff
@@ -158,11 +159,28 @@ class ScriptTile : public ScriptObject {
...
+	 * @return True if and only if the tile is a river tile.
+	 */
+	static bool IsRiverTile(TileIndex tile);
+
 	/**
 	 * Checks whether the given tile is actually a water tile.
 	 * @param tile The tile to check on.
 	 * @pre ScriptMap::IsValidTile(tile).
 	 * @return True if and only if the tile is a water tile.
+	 * @note Returns false when a buoy was placed on the tile.
```
Comment (TrueBrain) [inline src/script/api/script_tile.hpp]: Nitpick, but "was placed" is wrong tense here. `Returns false when a buoy is on the tile` would be more clear, I think? Mind fixing that real quick? :)

**Review by TrueBrain — APPROVED** [review]

File: src/script/api/script_tile.hpp (hunk 2 above)
Comment (FLHerne) [reply src/script/api/script_tile.hpp]: What does it return for various other water-like tiles?
 * Ship depots (passable by ships)
 * Water-based industries and objects (impassable)

I was going to complain that this was an unintuitive exception before realising the function already existed.

Do the newly-added `IsSeaTile`/`IsRiverTile` have the same exception?  (looks like the answer is 'yes') If so, should probably note it there too.

There doesn't seem to be a function that checks for buoys, nor for 'passable water tiles' in general; would adding one be useful for AIs? I guess it would need a direction argument to make sense for coasts/depots and then be too complicated.

Comment (SamuXarick) [reply src/script/api/script_tile.hpp]: It's hard to answer this without consulting the results. There are test cases in regression that checks the tiles posted in the screenshot above. Rivers weren't tested, the map doesn't contain any river. Oil Rig wasn't tested either.

Item | IsWaterTile | IsSeaTile | IsRiverTile | IsCanalTile | IsCoastTile
---- | ----------- | ---------- | ----------- | ----------- | -----------
Ship Depot | yes | no | no | no | no
Buoy | no | no | no | no | no | no
Sea Tile | yes | yes | no | no | no
Lock Lower Part | yes | no | no | no | no
Lock Middle Part | yes | no | no | no | no
Lock Upper Part | yes | no | no | no | no
Coast Half-Tile | no | no | no | no | yes
Coast Full-Tile | no | no | no | no | yes
Canal Tile | yes | no | no | yes | no
Grass/Tree Ground Tile | no | no | no | no | no
Dock on Water | no | no | no | no | no
Dock on Slope | no | no | no | no | no

Comment (FLHerne) [reply src/script/api/script_tile.hpp]: Thanks for the detailed table!

Looking at that, I honestly think it would be better to return `true` for buoys in IsWaterTile and consider the previous behaviour as a bug instead of documenting it. It makes no sense for ship depots to qualify but not buoys.

I might make a separate PR.
